CARDS-1484 / CARDS-1349: Every night, query Torch to obtain patients who have appointments in the next 72h

OrdiNeu opened the pull request:
[Jira link for CARDS-1484](https://phenotips.atlassian.net/browse/CARDS-1484) 

[Jira link for CARDS-1349](https://phenotips.atlassian.net/browse/CARDS-1349) 

This PR introduces a new endpoint for importing data from the Torch server. The endpoint can be used for cronjobs if the `NIGHTLY_IMPORT_SCHEDULE` environment variable is set, similar to the data export. A manual import can be triggered by logging in as `admin` and accessing `<server url>/Subjects.import`. Any existing data should be overridden with what is on the Torch side when this occurs.

**The test branch is no longer required**. Instead, setup the OSGI config to query a sufficient number of days (15-30 for best results) as the test server has quite a few appointments in it now.

**The following OSGI configuratino must be set before testing!** Navigate to `http://localhost:<port>/system/console/configMgr` and go to the bottom of the list where there are two Import Configs (one is being used for the endpoint, the other is for the nightly imports). In both cases, set the Vault token to the JWT token in the last command by Pouria in \#datalake-endpoints on Slack.

**NEW! Also to be tested:** The provider IDs can now be specified from the OSGi config above. If the field is empty, no provider filtering is applied. A pipe `|` delimited list can be placed inside, and only appointments from providers matching those EPR IDs will be retrieved. For reference, a sample provider EPR ID is `135654`. In addition, in each `Visit Information` form:
1. Its "Practitioner" field should include the practitioner's full name including prefixes/suffixes (I don't know if there are any in the test data)
2. Its "location" field should be the nicer displayName as put out in the `/Proms/ClinicMapping/` folder.

OrdiNeu committed:
CARDS-1484: Every night, query Torch to obtain patients who have
appointments in the next 72h

Add the import task

OrdiNeu committed:
[wip] CARDS-1389 work

OrdiNeu committed:
CARDS-1349 Store patient and visit data locally

OrdiNeu committed:
CARDS-1484/CARDS-1349: Store patient and visit data locally

Fix up errors after merging with `dev`

OrdiNeu committed:
CARDS-1484/CARDS-1349: Store patient and visit data locally

Code cleanup

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Code cleanup and refactor

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Code cleanup

marta- commented on proms-resources/backend/pom.xml:
```diff
@@ -0,0 +1,90 @@
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.api</artifactId>
+      <version>2.18.4</version>
```
Let's follow the trend introduced by PR #818 and avoid specifying dependency versions in module `pom` files.

marta- commented on proms-resources/backend/pom.xml:
```diff
@@ -0,0 +1,90 @@
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.api</artifactId>
+      <version>2.18.4</version>
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>org.osgi.service.component.annotations</artifactId>
+      <version>1.4.0</version>
+    </dependency>
+    <dependency>
+      <groupId>javax.jcr</groupId>
+      <artifactId>jcr</artifactId>
+      <version>2.0</version>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.servlets.annotations</artifactId>
+      <version>1.2.4</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.commons.scheduler</artifactId>
+      <version>2.7.2</version>
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>org.osgi.service.component</artifactId>
+      <version>1.4.0</version>
+    </dependency>
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <version>3.1.0</version>
+      <scope>provided</scope>
+    </dependency>
+  </dependencies>
```
Let's follow the trend introduced by PR #818 and avoid specifying dependency versions in module `pom` files. 
```suggestion
  <dependencies>
    <dependency>
      <groupId>javax.json</groupId>
      <artifactId>javax.json-api</artifactId>
    </dependency>
    <dependency>
      <groupId>org.apache.sling</groupId>
      <artifactId>org.apache.sling.api</artifactId>
    </dependency>
    <dependency>
      <groupId>org.osgi</groupId>
      <artifactId>org.osgi.service.component.annotations</artifactId>
    </dependency>
    <dependency>
      <groupId>javax.jcr</groupId>
      <artifactId>jcr</artifactId>
    </dependency>
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-api</artifactId>
    </dependency>
    <dependency>
      <groupId>org.apache.sling</groupId>
      <artifactId>org.apache.sling.servlets.annotations</artifactId>
    </dependency>
    <dependency>
      <groupId>org.apache.sling</groupId>
      <artifactId>org.apache.sling.commons.scheduler</artifactId>
    </dependency>
    <dependency>
      <groupId>org.osgi</groupId>
      <artifactId>org.osgi.service.component</artifactId>
    </dependency>
    <dependency>
      <groupId>javax.servlet</groupId>
      <artifactId>javax.servlet-api</artifactId>
      <scope>provided</scope>
    </dependency>
  </dependencies>
```

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportEndpoint.java:
```diff
@@ -0,0 +1,62 @@
+
+        //Ensure that this can only be run when logged in as admin
+        final String remoteUser = request.getRemoteUser();
+        if (remoteUser == null || !"admin".equals(remoteUser)) {
```
Checking if `null` seems unnecessary here.
```suggestion
        if (!"admin".equals(remoteUser)) {
```

OrdiNeu replied:
Per [HttpServletRequests](https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getRemoteUser()), `remoteUser` can be `null` if the user is not logged in.

marta- replied:
But that's covered by `!"admin".equals(remoteUser)`.

OrdiNeu replied:
Added in `c58d36f3` :heavy_check_mark:

marta- replied:
Doesn't seem to have been pushed yet :)

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,416 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
```
Can this be avoided by refactoring to place the query logic and the local storage logic in separate classes?

OrdiNeu replied:
This warning appears to be unnecessary now after I refactored it. I've removed it in `82167d` :heavy_check_mark:

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,416 @@
+    /** Default log. */
+    private static final Logger LOGGER = LoggerFactory.getLogger(ImportTask.class);
+
+    private static final String URL = "https://prom.dev.uhn.io/graphql";
+
+    private static final String JWT_URL = "https://vault.dev.uhn.io/v1/auth/jwt/login";
```
These need to be taken from config file(s). we'll have different URLs in production.

OrdiNeu replied:
I've placed it into environment variables in `644b5`, to simulate how our endpoint does it :heavy_check_mark:

marta- replied:
Per @sdumitriu 's suggestion, OSGI config should be used instead.

OrdiNeu replied:
Applied in `f0acd29` :heavy_check_mark:

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,416 @@
+    private static final String JCR_SQL = "JCR-SQL2";
+    private static final String QUESTION_FIELD = "question";
+    private static final String VALUE_FIELD = "value";
+    private static final int DAYS_TO_QUERY = 3;
```
I'd make this a configuration as well.

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Remove unnecessary warning suppression

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Refactor the endpoint to use environment
variables

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Remove useless null check

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Switch to using OSGI config instead of
environment variables

OrdiNeu committed:
CARDS-1484 / CARDES-1349: Remove unused provisioning file

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportConfig.java:
```diff
@@ -0,0 +1,58 @@
+@ObjectClassDefinition(name = "Import config", description = "Configuration for the PROMs importer")
+public @interface ImportConfig
+{
+    /** Number of days to look ahead when querying for appointments. */
+    int DAYS_TO_QUERY = 3;
+
+    /** Torch FIHR GraphQL endpoint. */
+    String TORCH_ENDPOINT_URL = "https://prom.dev.uhn.io/graphql";
+
+    /** Vault JWT refresh endpoint. */
+    String VAULT_AUTH_URL = "https://vault.dev.uhn.io/v1/auth/jwt/login";
+
+    /** Cron-readable import schedule. */
+    String NIGHTLY_IMPORT_SCHEDULE = "";
+
+    /** Vault JWT token. */
+    String VAULT_TOKEN = "";
```
Can these be changed into `final` types if they are constants?

OrdiNeu replied:
Interface variables are both static and final by default -- the codestyler complains about redundant modifiers if I include them.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+        String postRequest = "{ \"role\": \"prom_role\", \"jwt\":\"" + this.vaultToken + "\" }";
+
+        try {
+            String rawResponse = getPostResponse(this.authURL, postRequest);
```
I'm not completely sure what is happening in this `getAuthToken()` method. It looks like a HTTP POST is made to `this.authURL` to validate `this.vaultToken` although this behavior is not implied by the function's name or its Javadoc. The output from `getPostResponse` is stored in the String variable `rawResponse` but it is never subsequently used.

OrdiNeu replied:
rawResponse is mostly captured for testing; the main point is that the POST request is made to Vault with our JWT so that we are temporarily authorized to access the Torch server. I'll rename this method to make it clearer. :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+        }
+        http.setDoOutput(true);
+
+        // Get the next date
```
Does using this method have anything specific to do with _dates_? Looking at the code and Javadoc for this method, it looks like this method POSTs any generic _string_ data that is not necessarily a date.

OrdiNeu replied:
Whoops, I think I moved some code around and this comment got moved too. Fixed in `ccec843` :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+
+    /**
+     * Grab a form of the specified type, or create it if it doesn't exist.
+     * @param patient Resource of the subject node for the form
```
This Javadoc should be fixed; `@param patient` should be changed to `@param subject` and the `String questionnairePath` parameter should be documented as well.

OrdiNeu replied:
Applied (with a few other comment fixes) in `ca03fd32` :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+
+    /**
+     * Store information about a patient from a JsonObject.
+     * @param patient The patient object to add
```
Change `@param patient` to `@param patientDetails` and specify type of `JsonObject`.

OrdiNeu replied:
The refactor in `9f1684c` has removed this entirely :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+
+    /**
+     * Get any upcoming appointments in the next few days.
+     * @param authToken an authentication token for use in querying the Torch server
```
The `daysToQuery` parameter should be documented as well.

OrdiNeu replied:
Applied in `ccec843` :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+     * @param resolver ResourceResolver to use when searching for/creating the node
+     * @return A Subject resource
+     */
+    Resource getOrCreateSubject(String identifier, String parentTypePath, Resource parent, ResourceResolver resolver)
```
In `modules/data-entry/src/main/java/io/uhndata/cards/DataImportServlet.java` the function `getOrCreateSubject` calls the function `createSubject` which performs a JCR check-in of the newly created _Subject_ node. Should that be done here as well? @sdumitriu

OrdiNeu replied:
Applied in `9f1684c` :heavy_check_mark:

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,433 @@
+     * @param resolver ResourceResolver to use when searching for/creating the node
+     * @return A Form resource
+     */
+    Resource getOrCreateForm(Resource subject, String questionnairePath, ResourceResolver resolver)
```
In `modules/data-entry/src/main/java/io/uhndata/cards/DataImportServlet.java` the function `parseRow` performs a JCR check-in of all the newly added _Form_ nodes after calling `getOrCreateForm`. Should something like that be done here as well? @sdumitriu

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportConfig.java:
```diff
@@ -0,0 +1,58 @@
+import org.osgi.service.metatype.annotations.AttributeType;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+
+@ObjectClassDefinition(name = "Import config", description = "Configuration for the PROMs importer")
```
I suggest splitting this file into two new files - `NightlyImportConfig.java` and `EndpointImportConfig.java` as the `Import schedule` parameter does not make sense in the context of a data import triggered by calling an API endpoint.

Also, by splitting into two different files, it will be possible to distinguish between the two configuration settings in `/system/console/configMgr`. I suggest doing the following:

- `NightlyImportConfig.java` should have a `name` set to `Nightly PROMs Import Config` and a `description` set to `Configuration for the PROMs nightly importer`.
- `EndpointImportConfig.java` should have a `name` set to `Endpoint-triggered PROMs Import Config` and a `description` set to `Configuration for the PROMs endpoint-triggered importer`.

OrdiNeu replied:
Applied in `9f1684c` :heavy_check_mark:

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Move the import config into two config files
for their respective uses.
Add the versioning checkin to used nodes, and split the import task into
two classes since it was running into fan-out complexity errors

OrdiNeu committed:
CARDS-1484 / CARDS-1349 Apply styling changes suggested by
@IntegralProgrammer

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Apply code review suggestions by
@IntegralProgrammer

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -27,62 +27,43 @@
-
-    /** Sling property name for the question reference field on an Answer node. */
-    private static final String QUESTION_FIELD = "question";
+    /** Numnber of days to query. */
```
`Numnber` --> `Number`

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,329 @@
+            return formResource.next();
+        } else {
+            Resource parentResource = resolver.getResource("/Forms/");
+            Resource newSubject = resolver.create(parentResource, UUID.randomUUID().toString(), Map.of(
```
Should this be renamed to `newForm`?

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,329 @@
+    public void store()
+    {
+        String mrn = this.patientDetails.getString("mrn");
+        this.nodesToCheckin = new HashSet<String>();
```
This causes `this.nodesToCheckin` to be set to an empty (String) set and therefore when `this.nodesToCheckin.forEach` is called later in this function it will always be iterating over an empty set and therefore no nodes will actually be checked in.

OrdiNeu replied:
They are filled out during calls to `getOrCreateSubject` or `getOrCreateForm`. Would it be clearer if I pass around `nodesToCheckin` to those functions?

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Apply code review suggestions by
@IntegralProgrammer

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Change local nodesToCheckin so that it's easier to track

IntegralProgrammer commented:
There seems to be a difference between the way that the _patient charts_ are displayed for Patient Information / Visit Information entered manually versus Patient Information / Visit Information imported from the Torch server.

For example, here is the patient chart obtained from the data imported from the Torch server
![data_from_torch](https://user-images.githubusercontent.com/5349728/146276763-bec528df-5152-4bf4-a586-ba35cfe4ace8.png)

...and here is the same data created by manually entering it through the web interface
![manually_created_data](https://user-images.githubusercontent.com/5349728/146276977-7df3eec5-f471-494e-a12c-9a717ac7084a.png)

IntegralProgrammer commented:
Probably not within the scope of this PR but the _Time_ field in the _Visit information_ questionnaire will not accept any input.

IntegralProgrammer commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+                PatientLocalStorage.QUESTION_FIELD, resolver.getResource(parentQuestionnaire + "/"
+                    + entry.getKey()).adaptTo(Node.class),
+                PatientLocalStorage.VALUE_FIELD, safelyGetValue(entry.getValue(), info),
+                PatientLocalStorage.PRIMARY_TYPE, "cards:TextAnswer"
+                ));
+        }
+
+        // Finally, do the same with date fields, which have different parameters
+        for (Map.Entry<String, String> entry : dateFields.entrySet()) {
+            try {
+                Calendar entryDate = Calendar.getInstance();
+                entryDate.setTime(new SimpleDateFormat("yyyy-MM-dd").parse(info.getString(entry.getValue())));
+                if (dateNodes.containsKey(entry.getKey())) {
+                    dateNodes.get(entry.getKey()).setProperty(PatientLocalStorage.VALUE_FIELD, entryDate);
+                } else {
+                    resolver.create(form, UUID.randomUUID().toString(), Map.of(
+                        PatientLocalStorage.QUESTION_FIELD,
+                        resolver.getResource(parentQuestionnaire + "/" + entry.getKey()).adaptTo(Node.class),
+                        PatientLocalStorage.VALUE_FIELD, entryDate,
+                        PatientLocalStorage.PRIMARY_TYPE, "cards:DateAnswer"
```
Should we support more data types than `cards:TextAnswer` and `cards:DateAnswer`?

sashaandjic commented:
> There seems to be a difference between the way that the _patient charts_ are displayed for Patient Information / Visit Information entered manually versus Patient Information / Visit Information imported from the Torch server.

Seems that import from the Torch server created **Visit information** form for the patient **7018120** rather than for the actual visit **AD51475YJ.E**. Is it possible that **Visit information** form was imported before visit **AD51475YJ.E** was created?

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Change the visit information subject to be the
visit instead of the patient

OrdiNeu commented:
> There seems to be a difference between the way that the _patient charts_ are displayed for Patient Information / Visit Information entered manually versus Patient Information / Visit Information imported from the Torch server.

Fixed in `099af34` :heavy_check_mark:

sdumitriu commented on proms-resources/backend/src/main/features/core/import-config.json:
```diff
@@ -0,0 +1,44 @@
+// under the License.
+
+{
+  "configurations":{
```
These 2 configurations look strikingly similar... Are you sure both are needed in their current form?

I think you should define the configuration only once, apply it to a tiny service that just exposes the configuration, and `@Reference` it in both importers. For increased security, that service must not be exported outside the bundle, so it should be an internal class (in a `.internal` java package).

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/EndpointImportConfig.java:
```diff
@@ -0,0 +1,54 @@
+    /** Number of days to look ahead when querying for appointments. */
+    int DAYS_TO_QUERY = 3;
+
+    /** Torch FIHR GraphQL endpoint. */
```
```suggestion
    /** Torch FHIR GraphQL endpoint. */
```

sdumitriu commented on proms-resources/backend/src/main/features/core/import-config.json:
```diff
@@ -0,0 +1,44 @@
+      // Vault JWT refresh endpoint.
+      "torch.endpoint.url": "https://prom.dev.uhn.io/graphql",
+      // Vault JWT refresh endpoint.
+      "vault.auth.url": "https://vault.dev.uhn.io/v1/auth/jwt/login",
```
Hm, these are a bit too specific for my taste. Are we sure we should encode these in the sources?

sdumitriu commented on proms-resources/backend/src/main/features/core/import-config.json:
```diff
@@ -0,0 +1,44 @@
+      // Vault JWT refresh endpoint.
+      "vault.auth.url": "https://vault.dev.uhn.io/v1/auth/jwt/login",
+      // Vault JWT token
+      "vault.token": ""
```
All these values are already specified as defaults in the configuration interface, I don't think they need to be du(tri?)plicated here as well.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+@Component(service = { Servlet.class })
+@SlingServletResourceTypes(
+    resourceTypes = { "cards/SubjectsHomepage" },
+    selectors = { "import" })
```
`import` seems to generic, I would name this `torchImport`.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public void doGet(final SlingHttpServletRequest request, final SlingHttpServletResponse response) throws IOException
```
This isn't very RESTful. A `GET` request should only retrieve existing data, without any changes performed on the server. I would use `POST` instead, even if it makes calling it a bit harder.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,194 @@
+    {
+        // Since we can't query more than one date at a time, query three dates
+        String postRequestTemplate = "{\"query\": \"query{"
+            + "patientsByDateAndClinic(clinic: \\\"PMH 8 Palliative Care Oncology Clinic\\\", date: \\\"%s\\\") {"
```
The name of the clinic should be configurable at some point, but that may wait until after the demo.

marta- replied:
Let's pull this clinic name out of the code and into a configuration please

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,194 @@
+        // Since we can't query more than one date at a time, query three dates
+        String postRequestTemplate = "{\"query\": \"query{"
+            + "patientsByDateAndClinic(clinic: \\\"PMH 8 Palliative Care Oncology Clinic\\\", date: \\\"%s\\\") {"
+            + "name {given family} sex mrn ohip dob emailOk com {email} "
```
Huh, we don't seem to take into account `emailOk`. Am I right that it indicates whether the patient agrees to receiving emails? Because if that's true, we should not be sending them emails, unless the specs of the project specifically say we should.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,194 @@
+        BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream(), "utf-8"));
+        String responseLine = null;
+        StringBuilder rawResponse = new StringBuilder();
+        while ((responseLine = br.readLine()) != null) {
```
Hopefully there won't ever be enough patients to cause an OOM.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/NightlyImport.java:
```diff
@@ -0,0 +1,69 @@
+        LOGGER.info("NightlyImport activating");
+        final String nightlyImportSchedule = config.nightly_import_schedule();
+        ScheduleOptions options = this.scheduler.EXPR(nightlyImportSchedule);
+        options.name("NightlyImport");
```
```suggestion
        options.name("NightlyTorchImport");
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,194 @@
+                JsonArray data = response.getJsonObject("data").getJsonArray("patientsByDateAndClinic");
+                for (int j = 0; j < data.size(); j++) {
+                    PatientLocalStorage thisPatient = new PatientLocalStorage(data.getJsonObject(j),
+                        this.resolverFactory.getResourceResolver(null));
```
This is strange, I would expect this to not work in the absence of credentials. I think a service user is needed, along with calling `this.resolverFactory.getServiceResourceResolver(null)` instead.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+public class PatientLocalStorage
+{
+    /** Default logger. */
+    private static final Logger LOGGER = LoggerFactory.getLogger(NightlyImport.class);
```
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(PatientLocalStorage.class);
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+     * @param nodesToCheckin Set of nodes that should be checked in after all edits are complete
+     * @return A Subject resource
+     */
+    Resource getOrCreateSubject(String identifier, String parentTypePath, Resource parent, ResourceResolver resolver,
```
Why is `resolver` passed around so much, since it's always the same as `this.resolver`?

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java:
```diff
@@ -0,0 +1,194 @@
+                    PatientLocalStorage thisPatient = new PatientLocalStorage(data.getJsonObject(j),
+                        this.resolverFactory.getResourceResolver(null));
+                    thisPatient.store();
+                }
```
The way `PatientLocalStorage` behaves seems fishy to me. A new object is created, then only one method can and must be called. It sounds like it wants to be service with one job, persisting patients, not data class.

I would rewrite this something like this:

```
final PatientLocalStorage storage = new PatientLocalStorage(this.resolverFactory.getServiceResourceResolver(null));
data.forEach(storage::store);
```

where `PatientLocalStorage` takes only the `ResourceResolver` parameter for the constructor, and has a `public void store(final JsonValue value)` function.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+        Set<String> nodesToCheckin)
+        throws RepositoryException, PersistenceException
+    {
+        Iterator<Resource> patientResource = resolver.findResources(String.format(
```
Why `patient`? Isn't this supposed to apply to both patients and subjects, thus a more generic `subject` is better suited?

The comment applies to both `patientResource` and `patientType`.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+    /**
+     * Grab a subject of the specified type, or create it if it doesn't exist.
+     * @param identifier Identifier to use for the subject
+     * @param parentTypePath path to a SubjectType node for this subject
```
Sounds more like `subjectTypePath` than `parent`, right?

```suggestion
     * @param subjectTypePath path to a SubjectType node for this subject
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+            Resource newSubject = resolver.create(parentResource, UUID.randomUUID().toString(), Map.of(
+                PatientLocalStorage.PRIMARY_TYPE, "cards:Subject",
+                "identifier", identifier,
+                "fullidentifier", identifier,
```
```suggestion
                "fullIdentifier", identifier,
```

But isn't this going to be computed automatically when saving anyway?

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+        Node subjectNode = subject.adaptTo(Node.class);
+        Iterator<Resource> formResource = resolver.findResources(String.format(
+            "SELECT * FROM [cards:Form] WHERE subject = \"%s\" AND questionnaire=\"%s\"",
+            subjectNode.getProperty("jcr:uuid").getString(),
+            formType.adaptTo(Node.class).getProperty("jcr:uuid").getString()), PatientLocalStorage.JCR_SQL);
```
It's recommended to use `.getIdentifier()` instead of `getProperty("jcr:uuid").getString()`.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+            Resource newForm = resolver.create(parentResource, UUID.randomUUID().toString(), Map.of(
+                PatientLocalStorage.PRIMARY_TYPE, "cards:Form",
+                "questionnaire", formType.adaptTo(Node.class),
+                "relatedSubjects", subjectNode,
```
Just like `fullIdentifier`, isn't this computed automatically? If not, it should always be an array, and it should contain both patient and visit when the form applies to a visit.

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -0,0 +1,332 @@
+     * @return The string, or an empty string if it does not exist
+     */
+    void updateForm(Resource form, JsonObject info, String parentQuestionnaire,
+        Map<String, JsonStringGetter> mapping, Map<String, String> dateFields, ResourceResolver resolver)
```
Why not make `dateFields` also a map of json getters? It's more flexible.

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Stylistic changes suggested by @sergiu-

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Adjust the appointment finder to find all
upcoming appointments instead of just the first

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Add the emailOk field to the list of fields
tracked

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Make a date getter for parsing dates

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Apply code review suggestions by @sdumitriu-

OrdiNeu committed:
CARDS-1484 / CARDS-1349 Remove the constant resolver passaround

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Refactoring suggestions by @sdumitriu-

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Apply service user and forEach optimization

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Pull out clinic name into an OSGi variable

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Support multiple clinics via a pipe-delimited
list

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Missing import

marta- commented:
In light of the latest information regarding how clinics are managed, this code needs to be changed as follows:
* We need to maintain a mapping from `clinic name` to `survey id`; multiple slightly different names may be used for the same clinic, and they should all be mapped to the same survey. This mapping will grow as we onboard more clinics. The initial version of the mapping, for testing, will be 
```
"PM-8 Palliative Care Oncology" -> "Cardio"
"PMH 8 Palliative Care Oncology Clinic" -> "Cardio"
```
* When the "Visit information" form is created, populate the `surveys` answer with the value that the visit's `location` is mapped to.

I see the supported clinics are now stored as a `|` delimited string. This will probably have to change to accommodate a possibly large number of value pairs.

marta- commented:
Also, this PR needs another rebase.

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Update to the new API endpoint, and handle
clinic name mapping

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Fix the clinic mapping reference, code cleanup,
and move nodesToCheckin to a class member.

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Add FIHR ID and location to forms

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Fix the version management checkin failures

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Allow filtering by provider ID

OrdiNeu commented:
> In light of the latest information regarding how clinics are managed, this code needs to be changed as follows:

This has now been complete. Also, the provider restriction info discussed in today's standup has also been implemented.

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Use eID instead of FHIR ID for provider query
filtering

OrdiNeu committed:
CARDS-1484 / CARDS-1349: Add the display name to clinic mapping, and
allow for the practitioner's full name to be used

sdumitriu committed:
CARDS-1484 / CARDS-1349
Unified the configuration, removed redundant configuration file, made admin detection case insensitive

sdumitriu committed:
CARDS-1484: Every night, query Torch to obtain patients who have appointments in the next 72h
- Changed the configuration from single instance to multi-instance factory, made the scheduler more dynamic
- Remove very specific configuration values from the default
- Added a default value for the scheduler expression
- Changed the clinic names and provider names configuration to be an actual array of values instead of a PIPE delimited single string
- Changed the import servlet to require specifying which configuration to import
- Changed the import servlet to return the status as JSON

sdumitriu committed:
CARDS-1484: Every night, query Torch to obtain patients who have appointments in the next 72h
Bugfix: provider name was not properly retrieved

sdumitriu committed:
[cleanup] Improved codestyle

sdumitriu committed:
[cleanup] Improved codestyle
